micro-rdk-installer/src/nvs/request.rs

```diff
@@ -0,0 +1,146 @@
+    // requires storage data to have already been populated with this
+    // information from the robot's app config json
+    let robot_id = storage_data.get_robot_id()?;
+    let robot_secret = storage_data.get_robot_secret()?;
+    let app_address = storage_data.get_app_address()?;
+
+    let creds = RPCCredentials::new(
+        Some(robot_id.clone()),
+        "robot-secret".to_string(),
+        robot_secret.clone(),
+    );
+    let agent = AgentInfo {
+        os: "esp32-build".to_string(),
+        host: gethostname::gethostname().to_str().unwrap().to_string(),
+        ips: vec![local_ip().unwrap().to_string()],
+        version: "0.0.1".to_string(),
+        git_revision: "".to_string(),
+        platform: Some("esp32-build".to_string()),
+    };
+    let cfg_req = ConfigRequest {
+        agent_info: Some(agent),
+        id: robot_id.clone(),
+    };
+    let dial = DialOptions::builder()
+        .uri(app_address.clone().as_str())
+        .with_credentials(creds)
+        .disable_webrtc()
+        .connect()
+        .await?;
+    let mut app_service = RobotServiceClient::new(dial);
+    let cfg = app_service.config(cfg_req).await?.into_inner();
+    let robot_config = cfg
+        .config
+        .ok_or(anyhow::Error::msg("no config for robot"))?;
+    let cloud_cfg = robot_config.cloud.unwrap_or_default();
+    storage_data.robot_credentials.robot_name = Some(
+        cloud_cfg
+            .local_fqdn
+            .split('.')
+            .next()
+            .unwrap_or("")
+            .to_string(),
+    );
+    storage_data.robot_credentials.local_fqdn =
+        Some(cloud_cfg.local_fqdn.replace('.', "-").to_string());
+    storage_data.robot_credentials.fqdn = Some(cloud_cfg.fqdn.replace('.', "-").to_string());
+    Ok(())
+}
+
+async fn read_certificates(storage_data: &mut ViamFlashStorageData) -> anyhow::Result<()> {
+    let robot_id = storage_data.get_robot_id()?;
+    let robot_secret = storage_data.get_robot_secret()?;
+    let app_address = storage_data.get_app_address()?;
+
+    let creds = RPCCredentials::new(
+        Some(robot_id.clone()),
+        "robot-secret".to_string(),
+        robot_secret.clone(),
+    );
+    let cert_req = CertificateRequest {
+        id: robot_id.clone(),
+    };
+    let dial = DialOptions::builder()
+        .uri(app_address.clone().as_str())
+        .with_credentials(creds)
+        .disable_webrtc()
+        .connect()
+        .await?;
+    let mut app_service = RobotServiceClient::new(dial);
+    let certs = app_service.certificate(cert_req).await?.into_inner();
+    let tls_cert = &certs
+        .tls_certificate
+        .split_inclusive("----END CERTIFICATE-----");
+    let srv_cert = &mut tls_cert.clone().take(2).collect::<String>();
+    srv_cert.push('\0');
+    let ca_cert = &mut tls_cert
+        .clone()
+        .take(1)
+        .map(der::Document::from_pem)
+        .filter(|s| s.is_ok())
+        .map(|s| s.unwrap().1.to_vec())
```

Comment:
might be interested in [`filter_map()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map)